File: 06_Linked_List/03_LL_Cycle_Detection/main.py
# Definition for singly-linked list node
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def hasCycle(self, head: ListNode) -> bool:
        if not head or not head.next:
            return False
        
        slow = head
        fast = head.next
        
        while fast != slow:
            if not fast or not fast.next:
                return False
            slow = slow.next
            fast = fast.next.next
        
        return True

# Helper function to create a linked list with a cycle
def create_linked_list_with_cycle(values, index):
    if not values:
        return None
    
    # Create nodes
    nodes = [ListNode(val) for val in values]
    
    # Link nodes
    for i in range(len(nodes) - 1):
        nodes[i].next = nodes[i + 1]
    
    # Create cycle if index is valid
    if index >= 0 and index < len(nodes):
        nodes[-1].next = nodes[index]
    
    return nodes[0] if nodes else None

# Main function to test the solution
if __name__ == "__main__":
    solution = Solution()
    
    # Example 1: [1,2,3,4], index = 1 (cycle at index 1)
    values1 = [1, 2, 3, 4]
    index1 = 1
    head1 = create_linked_list_with_cycle(values1, index1)
    has_cycle1 = solution.hasCycle(head1)
    print(f"Example 1: {has_cycle1}")  # Output: True
    
    # Example 2: [1,2], index = -1 (no cycle)
    values2 = [1, 2]
    index2 = -1
    head2 = create_linked_list_with_cycle(values2, index2)
    has_cycle2 = solution.hasCycle(head2)
    print(f"Example 2: {has_cycle2}")  # Output: False

File: 06_Linked_List/03_LL_Cycle_Detection/main.c
#include <stdio.h>
#include <stdlib.h>

// Definition for singly-linked list node
struct ListNode {
    int val;
    struct ListNode* next;
};

// Function to create a new node
struct ListNode* createNode(int val) {
    struct ListNode* newNode = (struct ListNode*)malloc(sizeof(struct ListNode));
    if (newNode == NULL) {
        perror("Memory allocation failed");
        exit(EXIT_FAILURE);
    }
    newNode->val = val;
    newNode->next = NULL;
    return newNode;
}

// Function to free the memory allocated for the linked list nodes
void freeList(struct ListNode* head) {
    struct ListNode* tmp;
    while (head != NULL) {
        tmp = head;
        head = head->next;
        free(tmp);
    }
}

// Function to detect cycle in a linked list
int hasCycle(struct ListNode* head) {
    if (head == NULL || head->next == NULL) {
        return 0; // No cycle if head is NULL or there's only one node
    }
    
    struct ListNode* slow = head;
    struct ListNode* fast = head->next;
    
    while (fast != slow) {
        if (fast == NULL || fast->next == NULL) {
            return 0; // No cycle if fast pointer reaches end of list
        }
        slow = slow->next;
        fast = fast->next->next;
    }
    
    return 1; // Cycle detected
}

// Helper function to create a linked list with a cycle
struct ListNode* createLinkedListWithCycle(int* values, int length, int index) {
    if (values == NULL || length <= 0) {
        return NULL;
    }
    
    struct ListNode* head = createNode(values[0]);
    struct ListNode* tail = head;
    struct ListNode* cycleStart = NULL;
    
    // Create the linked list
    for (int i = 1; i < length; ++i) {
        tail->next = createNode(values[i]);
        tail = tail->next;
        if (i == index) {
            cycleStart = tail;
        }
    }
    
    // Create cycle if index is valid
    if (cycleStart != NULL) {
        tail->next = cycleStart;
    }
    
    return head;
}

// Main function to test the solution
int main() {
    // Example 1: [1,2,3,4], index = 1 (cycle at index 1)
    int values1[] = {1, 2, 3, 4};
    int length1 = sizeof(values1) / sizeof(values1[0]);
    int index1 = 1;
    struct ListNode* head1 = createLinkedListWithCycle(values1, length1, index1);
    int hasCycle1 = hasCycle(head1);
    printf("Example 1: %s\n", hasCycle1 ? "true" : "false");
    freeList(head1);

    // Example 2: [1,2], index = -1 (no cycle)
    int values2[] = {1, 2};
    int length2 = sizeof(values2) / sizeof(values2[0]);
    int index2 = -1;
    struct ListNode* head2 = createLinkedListWithCycle(values2, length2, index2);
    int hasCycle2 = hasCycle(head2);
    printf("Example 2: %s\n", hasCycle2 ? "true" : "false");
    freeList(head2);

    return 0;
}
